Add sticker sending to Telegram alarms - image of found Pokémon.

drewreece opened the pull request:
## Description
This is similar to pull request #163 but instead of images to Twitter this sends stickers to  Telegram messages. In the Telegram API stickers can be sent as image uploads or via file_id of existing stickers. I found a 2 part pack of original Pokémon & gathered the file_id's manually. The ID's are listed in a new file 'telegram_stickers.json'. It would be good to have the images checked for errors, some could potentially be out of sequence. 
The stickers are hosted on telegram.org & do not need to be added to this project (I don't know if this helps with potential copyright/ usage issues as the assets are outside of the codebase). 

Use 'send_sticker':'bool' in the Telegram settings to enable/ disable (off by default).

Sticker packs can be added to your Telegram client via these links (not required for this feature as far as I can tell).
https://web.telegram.org/#/im?tgaddr=tg:%2F%2Faddstickers%3Fset%3DPokemonGen1
https://web.telegram.org/#/im?tgaddr=tg:%2F%2Faddstickers%3Fset%3DPokemonGeneration1Part2

## How Has This Been Tested?
Latest PokemonGo-Map & PokeAlarm develop branches running on Raspberry Pi. Only tested using Telegram for alarms. Enabling & disabling in config works. Sends to web & iOS client.

## Screenshots (if appropriate):
Web client
![telegram-stickers](https://cloud.githubusercontent.com/assets/345286/18104950/71a8b444-6ef5-11e6-83d6-4620fe25331c.png)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Wiki Update
@TODO
This config setting needs explaining in the Wiki for the Telegram config, I can add that if this feature seems worth adding.

## Checklist:
- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

I do not understand the required code style, it 'kinda sorta' looks the same :)
I also do not understand how this project handles 'defaults', they seem to be defined all over the place i.e. hardcoded in '_defaults' and 'default[]', defined in various .json config files and passed via settings.get(). Please explain what I should change if my use of default values is wrong.

kvangent commented:
@drewreece What's your plan for poketops and pokelures?

drewreece commented:
I hadn't considered them however if suitable stickers can be found it can be added to the list but the logic will need to accommodate the fact the id is not from a Pokémon.

What needs to be shown? 
- A lured stop
- Gyms being taken by each team 
- Gyms leaving each team? 
- Gyms going empty? 
 
I think a sticker pack will need to be made for those, they seem a little too obscure to exist already.

kvangent commented:
@drewreece For slack, we link to one of two pngs located in the icon folder at the moment. 

I also think it might be better to include the dict of numbers to id#'s in the __init__ for the telegram module.

frojasdev commented:
@drewreece How can i check if the numer is the same of the image?, I'm trying implement but when eevee apears the image is of vaporeon, same problem with all. ex. Bellsporut = >Weepinbell

drewreece commented:
I'm just trying to recheck it. You basically have to send the sticker & via telepot.

You also need to setup a listener to receive the stickers to get the file_id - that is the only way I found to get the ID's. It's a pain to do all of them by hand after you have added the sticker pack. 

There is a stickerbot if you can stand doing it all in Telegram search for the 'fileid' bot.

I'll try to post a tool to make it easy to send them all & see the ID's of ones you send.

drewreece commented:
Sorry it looks like the stickers are really out of order. I must have stared at it too many times trying to get them correct.

This gist should be dropped into alarms/Telegram, add your bot-id & chat-id & run it.
https://gist.github.com/drewreece/8da044105935f15b5f3e2779789960a0

I was converting it to a python dict, that is why that file has a use_json() & a use_pyversion(), it's a mess but I hope it can help to get these rearranged.

frojasdev commented:
@drewreece I fixed it.

drewreece commented:
@frojasdev, thank you :) 
They made my brain hurt. 74 is also wrong so I'll update it to become…

`
"74": "BQADBAADuwADA6ZnAXgaugJm0yQlAg",
`

Geodude was a duplicate of Graveler.

hk2046 commented:
Hi Raichu,
Would you please help me to verify my following step is correct or not? thanks!

Step 1: Amend telegram_alarm.py below
===============================================================
#Setup Logging
import logging
log = logging.getLogger(__name__)

#Python modules

#Local modules
from ..alarm import Alarm
from ..utils import *

#External modules
import telepot
 
class Telegram_Alarm(Alarm):
 	
	#Gather settings and create alarm
	def __init__(self, settings):
                self.sticker = parse_boolean(settings.get('send_sticker', 'True'))
                #load sticker file_id list from json
                if self.sticker:
			with open(get_path('alarms/Telegram/telegram_stickers.json')) as file:
				self.stickerlist = json.load(file)
		self.bot_token = settings['bot_token']
 		self.connect()
 		self.chat_id = settings['chat_id']
		self.send_map = parse_boolean(settings.get('send_map', "True"))
		self.title = settings.get('title', "A wild <pkmn> has appeared!")
		self.body = settings.get('body', "<gmaps> \n Available until <24h_time> (<time_left>).")
		self.startup_message = settings.get('startup_message', "True")
		log.info("Telegram Alarm intialized.")
		if parse_boolean(self.startup_message):
			self.client.sendMessage(self.chat_id, 'PokeAlarm activated! We will alert this chat about pokemon.')
	
	#(Re)establishes Telegram connection
	def connect(self):
		self.client = telepot.Bot(self.bot_token) 
 		
	#Send Pokemon Info 
 	def pokemon_alert(self, pkinfo):
                if self.sticker:
			if info['id'] in self.stickerlist:
				stickerargs = {
					'chat_id': alert['chat_id'],
					'sticker': self.stickerlist[info['id']],
					'disable_notification': 'True'
				}
				try_sending(log, self.connect, 'Telegram', self.client.sendSticker, stickerargs)
				title = replace(self.title, pkinfo)
				body =  replace(self.body, pkinfo)
				
		args = {
			'chat_id': self.chat_id,
			'text': '<b>' + title + '</b> \n' + body,
			'parse_mode': 'HTML',
			'disable_web_page_preview': 'False',
		}
		
		try_sending(log, self.connect, "Telegram", self.client.sendMessage, args)
		if self.send_map is True:
			locargs = { 
				'chat_id': self.chat_id,
				'latitude':      pkinfo['lat'],
				'longitude':  pkinfo['lng'],
				'disable_notification': 'True'
			}
			try_sending(log, self.connect, "Telegram (loc)", self.client.sendLocation, locargs)
================================================================

Step 2: Create a telegram_stickers.py and place to alarms/telegram

Step 3: Create a test-send.py and place to alarms/telegram

Step 4: Open cmd and run pokealarm by using command : python runwebhook.py

Am i right? thanks!

kvangent commented:
Just gonna leave a list here for what we are waiting for to accept this:
1. Gym and Pokestop icons added as stickers (Should be able to upload them according to https://telegram.org/blog/stickers-revolution)
2. Move all of the sticker ids to a python dictionary inside the `__init__.py` for the Telegram module
3. Allow customizing per alert (pokemon, pokestops, gyms) for enabling or disabling stickers. If not specified from alert, should default to the general option (which if not specified generally is True)

drewreece commented:
@kvangent, for number 2 I've made the json list into a python dict.
stickerlist = {…}
I put that inside alarms/Telegram/telegram_stickers.py
I try to import it within alarms/Telegram/__init__.py and it always fails. I used
`from telegram_stickers import stickerlist`

I tried importing before Telegram_Alarm is imported & it fails. I also tried putting the entire dict into __init_.py & that also fails… 'NameError: global name 'stickerlist' is not defined'

Sorry I don't know what I am doing wrong, should I be creating another class for it? As you can see, I can hack, but don't know much beyond that.

kvangent commented:
Putting it in its own py file is probably better because it's easier to see instead of hiding in the init. Where are you attempting the import from?

drewreece commented:
I tried importing inside _ _init_ _.py in alarms/Telegram before the Telegram_Alarm is imported.

kvangent commented:
I would add the line `from .sticker_list import dict_name` to Local modules in Telegram_Alarm.

drewreece commented:
Yep, that is the only place I can get it to work :)

hk2046 commented:
Hello, may i know it is possible running now? thanks!

drewreece commented:
@hk2046 no

hk2046 commented:
Just for your information.
I can successful run the send sticker function but some setting of alarms.json can't run.
For example :       "title": "B 幫你搵到小精靈 #<pkmn> 快D去捉佢啦!!",
      "body": "佢依家係 <address> !!仲有時間:<time_left> 直至:<24h_time>",

drewreece commented:
@hk2046, I don't know what that means. Test it on the original branch & see if it works there, I don't think it is related to these changes.

hk2046 commented:
@drewreece below as my alarms.json setting.
I can run the "title", "body" before. However, i can't run "title" and"body" after change the setting of telegram_alarm

{
  "alarms": [
    {
      "active": "False",
      "type": "pushbullet",
      "api_key": "YOUR_API_KEY"
    },
    {
      "active": "False",
      "type": "slack",
      "api_key": "YOUR_API_KEY",
      "channel": "#general"
    },
    {
      "active": "False",
      "type": "twilio",
      "account_sid": "YOUR_API_KEY",
      "auth_token": "YOUR_AUTH_TOKEN",
      "from_number": "YOUR_FROM_NUM",
      "to_number": "YOUR_TO_NUM"
    },
    {
      "active": "True",
      "type": "telegram",
      "bot_token": "xxxxxxxxxxxxxxxx",
      "chat_id": "@xxxxxxxxxxxxxxxx",
      "title": "B 幫你搵到小精靈 #<pkmn> 快D去捉佢啦!!",
      "body": "佢依家係 <address> !!仲有時間:<time_left> 直至:<24h_time>",
      "send_map": "True"
    },
    {
      "active": "False",
      "type": "pushover",
      "app_token": "YOUR_APP_TOKEN",
      "user_key": "YOUR_USER_KEY"
    },
    {
      "active": "False",
      "type": "twitter",
      "access_token": "YOUR_ACCESS_TOKEN",
      "access_secret": "YOUR_ACCESS_SECRET",
      "consumer_key": "YOUR_CONSUMER_KEY",
      "consumer_secret": "YOUR_CONSUMER_SECRET"
    }
  ],
  "pokemon": {
    "Bulbasaur": "True",
    "Ivysaur": "True",
    "Venusaur": "True",
    "Charmander": "True",
    "Charmeleon": "True",
    "Charizard": "True",
    "Squirtle": "True",
    "Wartortle": "True",
    "Blastoise": "True",
    "Caterpie": "False",
    "Metapod": "False",
    "Butterfree": "True",
    "Weedle": "False",
    "Kakuna": "False",
    "Beedrill": "True",
    "Pidgey": "False",
    "Pidgeotto": "False",
    "Pidgeot": "True",
    "Rattata": "False",
    "Raticate": "False",
    "Spearow": "False",
    "Fearow": "False",
    "Ekans": "True",
    "Arbok": "True",
    "Pikachu": "True",
    "Raichu": "True",
    "Sandshrew": "False",
    "Sandslash": "True",
    "Nidoran♀": "False",
    "Nidorina": "False",
    "Nidoqueen": "True",
    "Nidoran♂": "False",
    "Nidorino": "False",
    "Nidoking": "True",
    "Clefairy": "False",
    "Clefable": "False",
    "Vulpix": "True",
    "Ninetales": "True",
    "Jigglypuff": "False",
    "Wigglytuff": "True",
    "Zubat": "False",
    "Golbat": "False",
    "Oddish": "False",
    "Gloom": "True",
    "Vileplume": "True",
    "Paras": "False",
    "Parasect": "False",
    "Venonat": "False",
    "Venomoth": "False",
    "Diglett": "False",
    "Dugtrio": "False",
    "Meowth": "False",
    "Persian": "False",
    "Psyduck": "False",
    "Golduck": "True",
    "Mankey": "False",
    "Primeape": "True",
    "Growlithe": "True",
    "Arcanine": "True",
    "Poliwag": "False",
    "Poliwhirl": "False",
    "Poliwrath": "True",
    "Abra": "False",
    "Kadabra": "False",
    "Alakazam": "False",
    "Machop": "True",
    "Machoke": "False",
    "Machamp": "True",
    "Bellsprout": "False",
    "Weepinbell": "False",
    "Victreebel": "True",
    "Tentacool": "False",
    "Tentacruel": "True",
    "Geodude": "False",
    "Graveler": "False",
    "Golem": "True",
    "Ponyta": "True",
    "Rapidash": "True",
    "Slowpoke": "False",
    "Slowbro": "False",
    "Magnemite": "False",
    "Magneton": "True",
    "Farfetch'd": "True",
    "Doduo": "False",
    "Dodrio": "False",
    "Seel": "False",
    "Dewgong": "False",
    "Grimer": "False",
    "Muk": "False",
    "Shellder": "False",
    "Cloyster": "True",
    "Gastly": "False",
    "Haunter": "True",
    "Gengar": "True",
    "Onix": "True",
    "Drowzee": "False",
    "Hypno": "True",
    "Krabby": "False",
    "Kingler": "False",
    "Voltorb": "False",
    "Electrode": "False",
    "Exeggcute": "False",
    "Exeggutor": "True",
    "Cubone": "False",
    "Marowak": "False",
    "Hitmonlee": "True",
    "Hitmonchan": "True",
    "Lickitung": "False",
    "Koffing": "False",
    "Weezing": "False",
    "Rhyhorn": "False",
    "Rhydon": "True",
    "Chansey": "False",
    "Tangela": "False",
    "Kangaskhan": "False",
    "Horsea": "False",
    "Seadra": "False",
    "Goldeen": "False",
    "Seaking": "False",
    "Staryu": "True",
    "Starmie": "False",
    "Mr. Mime": "False",
    "Scyther": "True",
    "Jynx": "True",
    "Electabuzz": "True",
    "Magmar": "True",
    "Pinsir": "False",
    "Tauros": "False",
    "Magikarp": "True",
    "Gyarados": "True",
    "Lapras": "False",
    "Ditto": "False",
    "Eevee": "True",
    "Vaporeon": "True",
    "Jolteon": "True",
    "Flareon": "True",
    "Porygon": "True",
    "Omanyte": "False",
    "Omastar": "False",
    "Kabuto": "False",
    "Kabutops": "False",
    "Aerodactyl": "True",
    "Snorlax": "False",
    "Articuno": "False",
    "Zapdos": "False",
    "Moltres": "False",
    "Dratini": "False",
    "Dragonair": "False",
    "Dragonite": "False",
    "Mewtwo": "False",
    "Mew": "False"
  }
}

kvangent commented:
@hk2046 I'm going to need to ask you to refrain from posting on PR's for support. This feature is still under development, and will be ready when its ready. Until then if you are unable to get it to work please stick to the main branch.

kvangent commented:
@drewreece Let us know if you need anything from us to get this finished.

drewreece commented:
@kvangent, this is what I have.

Sticker pack on Telegram - https://telegram.me/addstickers/gymstop
Telegram require 512px png's with transparency & a white outline, so I 'shopped the images - frankly the pokestop looks bad but I don't have time to find or make a better one at the moment (maybe next week). I think it really should indicate a 'lured' stop anyway. I also wonder if the sticker pack is tied to an account, if I close my account does it disappear? I don't know how Telegram handle that - ideally you can make your own pack & control it if required @kvangent.

I haven't seen a gym change hands yet so I don't know if the icon is correct, I'm not scanning a large area & I stop when not catching so I don't see many notifications from gyms (I don't recall any at all).

I don't know how to handle the disabling of stickers for stops & gyms, sorry. They are enabled globally via the 'send_stickers' argument. I still don't really understand _defaults and default[] and how they are managed from the alarms.json & passed around via set_alert() and send_alert(). I added settings for each type but my logic was just wrong so I have removed that.

If you want to add this feature please feel free to abandon the bad parts & redo it. I don't really know what I'm doing :(
I have other commitments until next week so please don't wait on me. I have pushed my latest that seems to work for me.

drewreece commented:
@kvangent if you don't want this pr please just close it. 

I don't understand how to make your requirements work for toggling the sending of stickers on gyms & stops. Maybe I didn't make myself clear, I need your help otherwise please just close & reject it.

fwiw I did tweak the stickers :)

![preview](https://cloud.githubusercontent.com/assets/345286/18497712/909c6c22-7a28-11e6-8554-c36fdcae977c.png)

xh0n9x commented:
Can anyone advise
How do i merge this with the IV & Moveset notification setting? Thanks

kvangent commented:
@drewreece Sorry I haven't been around to address your questions, I was out of the country for a week. I will take a look at your changes and give you some comments within  the next day or so.

drewreece commented:
@kvangent, I understand. Please take your time, I realise my code is sloppy & appreciate you looking at it.

kvangent commented on alarms/Telegram/telegram_alarm.py:
```diff
@@ -64,10 +70,28 @@ def set_alert(self, settings, default):
 		return alert
  		
 	#Send Alert to Telegram
  	def send_alert(self, alert, info):
```
add a parameter 'sticker=None' to the end of this. Allows us to specify a sticker_id

drewreece replied:
I don't understand this 'sticker' parameter. Doesn't it need to be self.sticker or something to work in the class/ object thing? Also where is it meant to go? Inside send_alert or at the end of set_alert?

I tried adding the other suggestions but it doesn't work yet so I'm struggling to work out what is not correct.

kvangent replied:
The idea is that the sticker will be optional - so if no sticker_id is included, it won't add it to the args. By defaulting it to None, we can check if it is None (with the comment below) and only add it if its not. `self.sticker` is only going to be a boolean to enable/disable the sticker sending (for people who don't want stickers).

kvangent commented on alarms/Telegram/telegram_alarm.py:
```diff
@@ -84,7 +108,7 @@ def send_alert(self, alert, info):
 				'disable_notification': 'True'
 			}
 			try_sending(log, self.connect, "Telegram (Loc)", self.client.sendLocation, locargs)
```
Above this line add something like: 
```
If sticker_id :
 args['sticker'] = sticker_id
```
If a sticker is specified, add it to the args.

kvangent commented on alarms/Telegram/telegram_alarm.py:
```diff
@@ -84,7 +108,7 @@ def send_alert(self, alert, info):
+
 	#Trigger an alert based on Pokemon info
 	def pokemon_alert(self, pokemon_info):
 		self.send_alert(self.pokemon, pokemon_info)
```
Change the alerts to look something like this:
```
if self.pokemon.stickers:
    self.send_alert(self.pokemon, pokemon_info, stickerlist[info['id'])
elif:
    self.send_alert(self.pokemon, pokemon_info)
```

If stickers are enabled, add the sticker to the alert call, otherwise leave it off. For gyms and pokestops do stickerlist['gym'] or stickerlist['pokestop']

kvangent commented on alarms/Telegram/telegram_stickers.py:
```diff
@@ -0,0 +1,153 @@
+stickerlist = {
+    '1': 'BQADBAADJwADA6ZnAU_NNRcf64d1Ag',
```
Lets add 'gym' and 'pokestop' to this - that way all the sticker ID's are in one place, and if people want to change to their down stickers there is only 1 file needed.

kvangent commented:
Merged in #248.
